source/1_engine/rex_engine/src/cmdline.cpp

```diff
@@ -0,0 +1,205 @@
+#include "rex_engine/cmdline.h"
+
+#include "rex_engine/cmd_args.h"
+#include "rex_engine/diagnostics/assert.h"
+#include "rex_engine/diagnostics/logging/log_macros.h"
+#include "rex_engine/log.h"
+#include "rex_std/cstring.h"
+#include "rex_std/vector.h"
+
+namespace rex
+{
+  namespace cmdline
+  {
+    struct ActiveArgument
+    {
+      rsl::string_view argument;
+      rsl::string_view value;
+    };
+
+    class CommandLineArguments
+    {
+    public:
+      explicit CommandLineArguments(rsl::string_view cmdLine)
+      {
+        // verify the auto generated command line arguments
+        REX_ASSERT_X(verify_args(g_command_line_args.data(), g_command_line_args.size()), "You have ambuguous command line arguments");
+
+        if(cmdLine.empty())
+        {
+          REX_LOG(LogEngine, "No command line arguments to parse");
+          return;
+        }
+
+        parse_cmd_line(rsl::string_view(cmdLine));
+      }
+
+      rsl::optional<rsl::string_view> get_argument(rsl::string_view arg)
+      {
+        for(ActiveArgument active_arg: m_arguments)
+        {
+          // early optimization that strincmp can't do
+          if(arg.length() != active_arg.argument.length())
+            continue;
+
+          if(rsl::strincmp(arg.data(), active_arg.argument.data(), arg.length()) == 0)
+          {
+            return active_arg.value;
+          }
+        }
+
+        return rsl::nullopt;
+      }
+
+    private:
+      void parse_cmd_line(rsl::string_view cmdLine)
+      {
+        // all we have to do is split the command line based with spaces
+        // there's no need to perform any heap allocation for that.
+        count_t start_pos = cmdLine.find_first_not_of(' ', 0); // it's possible, although unlikely, that we start with a space
+
+        // all executables should be called with at least 1 argument
+        // the first argument should always be the path to the executable itself
+        REX_ASSERT_X(start_pos != -1, "No arguments given to your exe, there should always be at least 1 arg.");
```

Comment:
Maybe also REX_VERIFY_X although less error prone? As this might get caught in a debug session first.

Reply:
I'm not sure what we'd win if we have a REX_VERIFY here?

Reply:
Yeah my point exactly, I'm just a little concerned of potential issues ( as in programming mistakes and not user input ) within the auto tests as asserts are only fired within debug and most of our tests should run in release.

Reply:
All of our tests should run in all configs imo

Reply:
Not sure if that's useful, tests make sure functionality is not broken in other places and final functionality is release no? So I'm not sure if we should waste resources ( as in hardware resources ) to test in debug as this might give a misguided view on things

Reply:
We should become more than once something is broken in 1 config and not in others

Reply:
But we can have release only tests

Reply:
But debug is caught while we develop ... but sure if you are convinced we need it. I just don't see a point in when this could benefit us. Unless we start developing cross platform, debug runs will just result in what we run locally

Reply:
not all things are caught while we develop, I already found issues in the boot tests that we didn't know about